solutions/usaco-1206.mdx

```diff
@@ -0,0 +1,119 @@
+---
+id: usaco-1206
+source: USACO Silver 2022 February
+title: Redistributing Gifts
+author: Chongtian Ma
+---
+
+[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)
+
+## Explanation
+
+Similar to the official analysis, I created a directed graph with edges from $i$ to $j$ if cow $i$ prefers gift $j$ more compared to his own. 
+This is because it is always more optimal for cow $i$ to receive one of these gifts. In other words, if cow $i$ is able to receive a gift where
+it has an outgoing edge, then the solution will always be more optimal. Remember to also add an edge between cow $i$ and $i$ in the case where
+it cannot get a more preferred gift.
+
+Then, I broke up the graph into strongly connected components. If $i$ is in a strongly connected component, this means that cow $i$ can receive
+all gifts in the strongly connected component through a series of reassignments and is guaranteed to have one or more preferred gift (including itself).
+This is because if cow $i$ and cow $j$ is in the same SCC, then it is guaranteed cow $i$ can receive gift $j$ through a series of traversals (proof by
+SCC definition). Obviously you only want to assign cow $i$ with a more preferred gift.
+## Implementation
+
+I used Kosaraju's algorithm to find all SCCs in the graph. At the end, I just assigned cow $i$ with the first gift in the SCC that it has preference to.
+Time complexity is $O(M)$ where $M$ is the number of edges as mentioned at the bottom of the official analysis.
+
+<LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+#define all(x) x.begin(), x.end()
+const int N = 505;
+
+// for kosaraju purposes
+vector<vector<int>> graph(N), revgraph(N);
```

Comment:
vector<int> graph[N], revgraph[N];

Also you should probably make a comment describing that revgraph is the reversed graph

Reply:
Well, its a essential part of kosaraju and I think the name revgraph is self implied. I'll still add it though.